events/UserDataCollectionEvent.php

```diff
@@ -0,0 +1,53 @@
+<?php
+/**
+ * @link https://www.humhub.org/
+ * @copyright Copyright (c) HumHub GmbH & Co. KG
+ * @license https://www.humhub.com/licences
+ */
+
+namespace humhub\modules\legal\events;
+
+use humhub\modules\comment\models\Comment;
+use humhub\modules\file\models\File;
+use humhub\modules\like\models\Like;
+use humhub\modules\post\models\Post;
+use humhub\modules\rest\definitions\CommentDefinitions;
+use humhub\modules\rest\definitions\FileDefinitions;
+use humhub\modules\rest\definitions\LikeDefinitions;
+use humhub\modules\rest\definitions\PostDefinitions;
+use humhub\modules\rest\definitions\UserDefinitions;
+use humhub\modules\user\events\UserEvent;
+
+class UserDataCollectionEvent extends UserEvent
```

Comment:
I think we still need support for binary files here. (i.e. user uploads)

What do you think if the event also creates a temp directory in runtime during init? Uploads can then also be stored here in a subdirectory. 

At the end, the ExportService can zip the temp directory and call an $event->cleanup, which deletes the temp directory again.

Reply:
@luke- Do you mean we should include in ZIP archive all files which were uploaded by the User?

Please note before create ZIP archive I don't create the json files on disk, the json files are append to ZIP archive from string:

```php
foreach ($this->getDataFiles() as $name => $fileContent) {
    $archive->addFromString('files/' . $name . '.json', json_encode($fileContent));
}
```

I don't think we should create a temp folder with the user's uploaded files before put them in ZIP archive, I think it can be done by code like this `$archive->addFile($realFilePath, $fileName)`.

Why do you want to move the logic from the `ExportService` into the `UserDataCollectionEvent`?
In order to make the service file smaller?
I thought it is ok to keep all logic inside service file or don't I understand something? :)

Reply:
> @luke- Do you mean we should include in ZIP archive all files which were uploaded by the User?
> 
Yes, we should also include all files, created_by the user.

> Please note before create ZIP archive I don't create the json files on disk, the json files are append to ZIP archive from string:
> 
> ```
> foreach ($this->getDataFiles() as $name => $fileContent) {
>     $archive->addFromString('files/' . $name . '.json', json_encode($fileContent));
> }
> ```
> 
Ok, thats fine!

> I don't think we should create a temp folder with the user's uploaded files before put them in ZIP archive, I think it can be done by code like this `$archive->addFile($realFilePath, $fileName)`.
> 
Ok, perfect. Then a temp folder is not required. But we could still create a method in the Event to collect file names.

> Why do you want to move the logic from the `ExportService` into the `UserDataCollectionEvent`? In order to make the service file smaller? I thought it is ok to keep all logic inside service file or don't I understand something? :)

Yes, my plan was to make the service a bit smaller. But I also wanted to separate the collection logic from the "control" service (download, create package, etc.)  Anyway, I like the current solution. Or what do you think? Was it a bad idea to create a own event class?

Reply:
@luke- Ok I have moved the creating of the ZIP package into the event, please review the [commit](https://github.com/humhub-contrib/legal/pull/77/commits/445058a081a9745607a5e1cc3560619e52908ae3).

> Was it a bad idea to create a own event class?

No, it is a good idea, I just didn't think we should keep many logic in event files, but if you think it is ok I have implemented such way, please review that, thanks for the idea ;)

Reply:
> @luke- Ok I have moved the creating of the ZIP package into the event, please review the [commit](https://github.com/humhub-contrib/legal/pull/77/commits/445058a081a9745607a5e1cc3560619e52908ae3).
> 
> > Was it a bad idea to create a own event class?
> 
> No, it is a good idea, I just didn't think we should keep many logic in event files, but if you think it is ok I have implemented such way, please review that, thanks for the idea ;)
> 
Yes, it's true, the event class should not have any business logic. But I would leave it as it is.

> Structure of the package:
> 
I think it is currently difficult to link the files from the ‘files.json’ to the upload if file names are duplicated, isn't it?


Would the following package structure be ok?
/files.json
/post.json 
/files/12424-2424ds-s2141-de/filename.jpeg

We would have a separate subdirectory for each upload/file, but the mapping is possible using the GUID.